cylc/flow/network/scan.py

```diff
@@ -13,302 +13,361 @@
 #
 # You should have received a copy of the GNU General Public License
 # along with this program.  If not, see <http://www.gnu.org/licenses/>.
-"""Port scan utilities."""
+"""Functionality for searching for workflows running as the current user.
 
+The :py:func:`scan` asynchronous generator yields workflows. Iterate
+over them using an ``async for`` statement::
+
+    async for flow in scan():
+        print(flow['name'])
+
+For further functionality construct a pipe::
+
+    pipe = scan | is_active(True) | contact_info
+    async for flow in pipe:
+        print(f'{flow["name"]} {flow["CYLC_SUITE_HOST"]}')
+
+There are filters which you can you to omit workflows e.g.
+:py:func:`cylc_version` and transformers which acquire more information
+e.g. :py:func:`contact_info`.
+
+.. note: we must manually list functions so they get built into the docs
+
+.. autofunction:: scan
+.. autofunction:: filter_name
+.. autofunction:: is_active
+.. autofunction:: contact_info
+.. autofunction:: cylc_version
+.. autofunction:: api_version
+.. autofunction:: graphql_query
+.. autofunction:: title
+
+"""
+
+from collections.abc import Iterable
 import asyncio
-import os
-from pwd import getpwall
+from pathlib import Path
 import re
-import sys
-import socket
-from cylc.flow import LOG
 
-from cylc.flow.exceptions import SuiteServiceFileError
-import cylc.flow.flags
-from cylc.flow.hostuserutil import is_remote_host, get_host_ip_by_name
+from pkg_resources import (
+    parse_requirements,
+    parse_version
+)
+
+from cylc.flow import LOG
+from cylc.flow.async_util import (
+    pipe,
+    asyncqgen,
+    scandir
+)
 from cylc.flow.network.client import (
     SuiteRuntimeClient, ClientError, ClientTimeout)
+from cylc.flow.exceptions import SuiteStopped
 from cylc.flow.platforms import platform_from_name
 from cylc.flow.suite_files import (
     ContactFileFields,
     SuiteFiles,
-    load_contact_file,
     get_suite_title,
-    get_suite_source_dir
+    load_contact_file_async
 )
 
-DEBUG_DELIM = '\n' + ' ' * 4
-INACTIVITY_TIMEOUT = 10.0
-MSG_QUIT = "QUIT"
-MSG_TIMEOUT = "TIMEOUT"
-SLEEP_INTERVAL = 0.01
 
+SERVICE = Path(SuiteFiles.Service.DIRNAME)
+CONTACT = Path(SuiteFiles.Service.CONTACT)
+SUITERC = Path(SuiteFiles.SUITE_RC)
 
-def async_map(coroutine, iterator):
-    """Map iterator iterator onto a coroutine.
 
-    * Yields results in order as and when they are ready.
-    * Slow workers can block.
+async def dir_is_flow(listing):
```

Comment:
Would this not be considered CPU-bound not IO-bound? I could be completely wrong here but I thought asyncio was for IO bound and processes were for CPU-bound?
So, if we wait for an IO operation it is some other machine/database doing the processing and we can free up our thread to use elsewhere while we wait. But if we wait for this method we're still burning our CPU anyway but also have the expense of setting up additional async. 
I don't think there is a bottleneck here that can be freed by adding async. 
I really don't feel super-confident on this yet, so might be making mistakes - sorry if that's the case! 
https://leimao.github.io/blog/Python-Concurrency-High-Level/

Reply:
Er, yeah, there is no IO in this function so the `async` is not necessary (there was earlier on in development). I think the `async` here is pretty harmless but I'll remove it anyway.

Note that the async transformer/filter functions which follow use `async` whether they do IO or not, simply because the interface is `async` (if one of them is `async` they all have to be), so the unnecessary `async def`s there should be ignored.